Documentation/ABI/JSON.md

```diff
@@ -0,0 +1,227 @@
+tests and, as the name suggests, configures the test run. The configuration is
+encoded as a single [JSON Lines](https://jsonlines.org) value.
+
+```
+<configuration-record> ::= {
+  <version>,
+  "kind": "configuration",
+  "payload": <configuration>
+}
+
+<configuration> ::= {
+  ["verbosity": <number>,] ; 0 is the default; higher means more verbose output
+                           ; while negative values mean quieter output.
+  ["filters": <array:test-filter>,] ; how to filter the tests in the test run
+  ["parallel": <bool>,] ; whether to enable parallel testing (on by default)
+  ; more TBD
+}
+
+<test-filter> ::= <test-filter-tag> | <test-filter-id>
+
+<test-filter-action> ::= "include" | "exclude"
+
+<test-filter-tag> ::= {
+  "action": <test-filter-action>,
+  "tags": <array:string>, ; the names of tags to include
+  "operator": <test-filter-tag-operator> ; how to combine the values in "tags"
+}
+
+<test-filter-tag-operator> ::= "any" | "all"
+
+<test-filter-id> ::= {
+  "action": <test-filter-action>,
+  "id": <test-id> ; the ID of the test to filter in/out
+}
+```
+-->
+
+### Streams
+
+A stream consists of a sequence of values encoded as [JSON Lines](https://jsonlines.org).
+A single instance of `<output-stream>` is defined per test process and can be
+accessed by passing `--experimental-event-stream-output` to the test executable
+created by `swift build --build-tests`.
+
+```
+<output-stream> ::= <output-record>\n | <output-record>\n <output-stream>
+```
+
+### Records
+
+Records represent the values produced on a stream. Each record is encoded on a
+single line and can be decoded independently of other lines. If a decoder
+encounters a record whose `"kind"` field is unrecognized, the decoder should
+ignore that line.
+
+```
+<output-record> ::= <metadata-record> | <test-record> | <event-record>
+
+<metadata-record> ::= {
+  <version>,
+  "kind": "metadata",
+  "payload": <metadata>
+}
+
+<test-record> ::= {
+  <version>,
+  "kind": "test",
+  "payload": <test>
+}
+
+<event-record> ::= {
+  <version>,
+  "kind": "event",
+  "payload": <event>
+}
+```
+
+### Metadata
+
+Metadata records are reserved for future use.
+
+```
+<metadata> ::= {
+  ; unspecified JSON object content
+}
+```
+
+### Tests
+
+Test records represent individual test functions and test suites. Test records
+are passed through the record stream **before** most events.
+
+<!--
+If a test record represents a parameterized test function whose inputs are
+enumerable and can be independently replayed, the test record will include an
+additional `"testCases"` field describing the individual test cases.
+-->
+
+```
+<test> ::= {
+  "kind": <test-kind>,
+  "name": <string>, ; the unformatted function or non-qualified type name
+  ["displayName": <string>,] ; the user-supplied custom display name
+  "sourceLocation": <source-location>, ; where the test is defined
+  "id": <test-id>,
+}
+
+<test-kind> ::= "suite" | "function" | "parameterizedFunction"
```

Comment:
I'm curious about rationale for parameterized tests being considered a distinct "kind"? Alternatively I could imagine them being represented as `test-kind ::= "function"`, with a distinct, optional Boolean property like `parameterized ::= true | false`

Reply:
No real rationale beyond "this is what I typed." If you feel strongly, I can refactor it.

Reply:
Refactored.